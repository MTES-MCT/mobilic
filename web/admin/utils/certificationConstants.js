/**
 * Certification constants
 * Ensures consistency between backend (English) and frontend (French) naming
 */

// TODO: After PR merge, uncomment these imports
// import {
//   CertificationBronze,
//   CertificationArgent,
//   CertificationOr,
//   CertificationDiamant
// } from "common/utils/icons";

const MEDAL_LEVELS_CONFIG = {
  BRONZE: { frontend: "BRONZE", display: "Bronze", filename: "bronze" },
  SILVER: { frontend: "ARGENT", display: "Argent", filename: "argent" },
  GOLD: { frontend: "OR", display: "Or", filename: "or" },
  DIAMOND: { frontend: "DIAMANT", display: "Diamant", filename: "diamant" }
};

export const BACKEND_MEDAL_LEVELS = Object.keys(MEDAL_LEVELS_CONFIG);
export const FRONTEND_MEDAL_LEVELS = Object.values(MEDAL_LEVELS_CONFIG).map(
  config => config.frontend
);

export const BACKEND_TO_FRONTEND_MAPPING = Object.entries(
  MEDAL_LEVELS_CONFIG
).reduce(
  (acc, [backend, config]) => ({ ...acc, [backend]: config.frontend }),
  {}
);

export const MEDAL_DISPLAY_LABELS = Object.entries(MEDAL_LEVELS_CONFIG).reduce(
  (acc, [, config]) => ({ ...acc, [config.frontend]: config.display }),
  {}
);

export const MEDAL_CDN_FILENAMES = Object.entries(MEDAL_LEVELS_CONFIG).reduce(
  (acc, [, config]) => ({ ...acc, [config.frontend]: config.filename }),
  {}
);

export const CDN_BASE_URL =
  "https://cdn.jsdelivr.net/gh/MTES-MCT/mobilic@master/common/assets/images/certification/squared";

// TODO: After PR merge, uncomment this mapping
// export const MEDAL_BADGE_COMPONENTS = {
//   BRONZE: CertificationBronze,
//   ARGENT: CertificationArgent,
//   OR: CertificationOr,
//   DIAMANT: CertificationDiamant,
//   SILVER: CertificationArgent,
//   GOLD: CertificationOr,
//   DIAMOND: CertificationDiamant
// };

/**
 * Internal utility to map a level to a specific mapping object
 * @param {string} level - Medal level (backend or frontend)
 * @param {Object} mapping - Target mapping object
 * @returns {string|null} Mapped value or null if level is invalid
 */
function getMappedValue(level, mapping) {
  if (!level) return null;
  const frontendLevel = BACKEND_TO_FRONTEND_MAPPING[level] || level;
  return mapping[frontendLevel] || null;
}

/**
 * Gets the display label for a medal level
 * @param {string} level - Medal level (backend or frontend)
 * @returns {string|null} Display label (Bronze, Argent, Or, Diamant) or null
 */
export function getMedalDisplayLabel(level) {
  return getMappedValue(level, MEDAL_DISPLAY_LABELS);
}

/**
 * Gets the CDN filename for a medal level
 * @param {string} level - Medal level (backend or frontend)
 * @returns {string|null} Filename (bronze, argent, or, diamant) or null
 */
export function getMedalCdnFilename(level) {
  return getMappedValue(level, MEDAL_CDN_FILENAMES);
}

/**
 * Gets the complete CDN URL for a medal badge
 * @param {string} level - Medal level (backend or frontend)
 * @returns {string|null} Complete CDN URL or null if level is invalid
 */
export function getMedalCdnUrl(level) {
  const filename = getMedalCdnFilename(level);
  if (!filename) return null;
  return `${CDN_BASE_URL}/${filename}.png`;
}

/**
 * Validates if a medal level is valid (backend or frontend)
 * @param {string} level - Medal level to validate
 * @returns {boolean} True if valid, false otherwise
 */
export function isValidMedalLevel(level) {
  return (
    BACKEND_MEDAL_LEVELS.includes(level) ||
    FRONTEND_MEDAL_LEVELS.includes(level)
  );
}
