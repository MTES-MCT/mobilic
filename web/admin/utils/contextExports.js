import React, { createContext, useContext } from "react";
import { useExports as useExportsHook } from "../hooks/useExports";

const ExportsContext = createContext(null);

export const useExportsContext = () => useContext(ExportsContext);

export function ExportsProvider({ children }) {
  return (
    <ExportsContext.Provider value={{ ...useExportsHook() }}>
      {children}
    </ExportsContext.Provider>
  );
}
