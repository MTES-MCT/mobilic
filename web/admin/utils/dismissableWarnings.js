export const DISMISSABLE_WARNINGS = {
  EMPLOYEE_MISSION_VALIDATION: "employee-validation",
  ADMIN_MISSION_MODIFICATION: "admin-mission-modification"
};
