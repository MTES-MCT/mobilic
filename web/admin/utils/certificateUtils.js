import { getMedalDisplayLabel, getMedalCdnUrl } from "./certificationConstants";

// ===== FONCTIONS D'INTERFACE PUBLIQUE =====
// Ces fonctions maintiennent la compatibilit√© avec l'API existante

/**
 * @deprecated Utiliser getMedalDisplayLabel de certificationConstants
 */
export function getCertificateLevelLabel(level) {
  return getMedalDisplayLabel(level);
}

/**
 * @deprecated Utiliser getMedalCdnUrl de certificationConstants
 */
export function getBadgeCdnUrl(level) {
  return getMedalCdnUrl(level);
}

/**
 * @deprecated Will be implemented after PR merge
 */
export function getCertificateBadgeComponent() {
  return null;
}

export function generateEmbedCodes(level) {
  const badgeUrl = getBadgeCdnUrl(level);
  const levelLabel = getCertificateLevelLabel(level);

  if (!badgeUrl) return { iframe: "", image: "", script: "" };

  const altText = `Certificat Mobilic ${levelLabel || level}`;
  const titleText = `Certificat Mobilic ${levelLabel || level}`;

  return {
    iframe: `<iframe src="${badgeUrl}" width="150" height="150" frameborder="0" style="border: none;" title="${titleText}"></iframe>`,

    image: `<img src="${badgeUrl}" alt="${altText}" width="150" height="150" />`,

    script: `<a href="https://mobilic.beta.gouv.fr" target="_blank" rel="noopener">
  <img src="${badgeUrl}" alt="${altText}" width="150" height="150" />
</a>`
  };
}
