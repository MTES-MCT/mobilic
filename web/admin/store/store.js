import React from "react";
import { useStoreSyncedWithLocalStorage } from "common/store/store";
import { adminRootReducer } from "./reducers/root";
import { getStartOfMonth, isoFormatLocalDate, now } from "common/utils/time";

export const VIRTUAL_ACTIVITIES_ACTIONS = {
  create: "create",
  edit: "edit",
  cancel: "cancel"
};

export const VIRTUAL_EXPENDITURES_ACTIONS = {
  create: "create",
  cancel: "cancel"
};

const AdminStoreContext = React.createContext(() => {});

export function AdminStoreProvider({ children }) {
  const store = useStoreSyncedWithLocalStorage();

  const [state, setState] = React.useState({
    companyId: undefined,
    users: [],
    workDays: [],
    vehicles: [],
    settings: [],
    knownAddresses: [],
    employments: [],
    companies: [],
    missions: [],
    minWorkDaysCursor: [],
    activitiesFilters: {
      period: "day",
      users: [],
      maxDate: isoFormatLocalDate(new Date(Date.now())),
      minDate: isoFormatLocalDate(getStartOfMonth(now()))
    },
    virtualActivities: [],
    virtualExpenditureActions: []
  });

  function dispatch(action) {
    setState(st => adminRootReducer(st, action));
  }

  return (
    <AdminStoreContext.Provider
      value={{
        ...state,
        userId: store.userId(),
        minWorkDaysDate: state.minWorkDaysCursor
          ? state.minWorkDaysCursor.slice(0, 10)
          : null,
        dispatch
      }}
    >
      {children}
    </AdminStoreContext.Provider>
  );
}

export const useAdminStore = () => React.useContext(AdminStoreContext);

export const useAdminCompanies = () => {
  const adminStore = useAdminStore();

  const [company, setCompany] = React.useState(null);
  const [sortedCompanies, setSortedCompanies] = React.useState([]);

  const companies = adminStore.companies;

  React.useEffect(() => {
    if (adminStore.companyId && companies && companies.length > 0) {
      setCompany(companies.find(c => c.id === adminStore.companyId));
      setSortedCompanies(
        companies.sort((c1, c2) =>
          c1.name.localeCompare(c2.name, undefined, {
            numeric: true,
            sensitivity: "base"
          })
        )
      );
    } else {
      setCompany(null);
    }
  }, [companies, adminStore.companyId]);

  return [sortedCompanies, company];
};
