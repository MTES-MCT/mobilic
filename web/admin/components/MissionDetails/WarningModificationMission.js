import React from "react";
import { Collapse, ListItemIcon, ListItemText, Alert } from "@mui/material";
import List from "@mui/material/List";
import ListItem from "@mui/material/ListItem";
import { Link } from "../../../common/LinkButton";
import { useStoreSyncedWithLocalStorage } from "common/store/store";
import { DISMISSABLE_WARNINGS } from "../../utils/dismissableWarnings";
import { useWarningModificationMissionStyles } from "./WarningModificationMissionStyle";
import { DISABLE_WARNING_MUTATION } from "common/utils/apiQueries";
import { captureSentryException } from "common/utils/sentry";
import { useApi } from "common/utils/api";

export function WarningModificationMission() {
  const store = useStoreSyncedWithLocalStorage();
  const userInfo = store.userInfo();
  const classes = useWarningModificationMissionStyles();
  const api = useApi();

  const [modificationAlertOpen, setModificationAlertOpen] = React.useState(
    !userInfo.disabledWarnings ||
      !userInfo.disabledWarnings.includes(
        DISMISSABLE_WARNINGS.ADMIN_MISSION_MODIFICATION
      )
  );

  async function handleDismissWarning() {
    try {
      await api.graphQlMutate(
        DISABLE_WARNING_MUTATION,
        { warningName: DISMISSABLE_WARNINGS.ADMIN_MISSION_MODIFICATION },
        { context: { nonPublicApi: true } }
      );
      await store.setUserInfo({
        ...store.userInfo(),
        disabledWarnings: userInfo.disabledWarnings.concat([
          DISMISSABLE_WARNINGS.ADMIN_MISSION_MODIFICATION
        ])
      });
      setModificationAlertOpen(false);
    } catch (err) {
      captureSentryException(err);
    }
  }

  return (
    <Collapse in={modificationAlertOpen}>
      <Alert
        key={0}
        severity="info"
        className={classes.modificationAlert}
        data-testid="warningAlert"
      >
        <List disablePadding>
          <ListItem
            disableGutters
            dense
            className={classes.modificationWarningItem}
          >
            <ListItemIcon className={classes.validationWarningIcon}>
              ðŸ‘‰
            </ListItemIcon>
            <ListItemText primary="Les modifications sont enregistrÃ©es et apparaÃ®tront en cas de contrÃ´le." />
          </ListItem>
          <ListItem
            disableGutters
            dense
            className={classes.modificationWarningItem}
          >
            <ListItemIcon className={classes.validationWarningIcon}>
              ðŸ‘‰
            </ListItemIcon>
            <ListItemText primary="En cas de modification des saisies, le salariÃ© recevra une notification." />
          </ListItem>
        </List>
        <Link
          className={classes.dismissModificationAlert}
          to="/#"
          data-testid="dismissMissionModificationWarningLink"
          onClick={e => {
            e.preventDefault();
            handleDismissWarning();
          }}
        >
          Ne plus afficher ce message
        </Link>
      </Alert>
    </Collapse>
  );
}
