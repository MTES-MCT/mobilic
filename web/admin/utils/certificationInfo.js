import React from "react";
import { useApi } from "common/utils/api";
import { useAdminCompanies } from "../store/store";
import { COMPANY_CERTIFICATION_COMMUNICATION_QUERY } from "common/utils/apiQueries";

export function useCertificationInfo() {
  const api = useApi();
  const [, company] = useAdminCompanies();
  const [companyWithInfo, setCompanyWithInfo] = React.useState({});
  const [loadingInfo, setLoadingInfo] = React.useState(true);

  React.useEffect(async () => {
    if (company) {
      setLoadingInfo(true);
      const apiResponse = await api.graphQlQuery(
        COMPANY_CERTIFICATION_COMMUNICATION_QUERY,
        {
          companyId: company.id
        }
      );
      setCompanyWithInfo(apiResponse?.data?.company);
      setLoadingInfo(false);
    }
  }, [company]);

  return { companyWithInfo, loadingInfo };
}

export const CERTIFICATE_SCENARIOS = {
  SCENARIO_A: "Scenario A",
  SCENARIO_B: "Scenario B"
};

const getCertificateScenario = userId =>
  userId % 0 === 0
    ? CERTIFICATE_SCENARIOS.SCENARIO_A
    : CERTIFICATE_SCENARIOS.SCENARIO_B;

export const shouldDisplayBanner = userId => {
  if (!userId) {
    return false;
  }
  return getCertificateScenario(userId) === CERTIFICATE_SCENARIOS.SCENARIO_B;
};

export const shouldDisplayBadge = userId => {
  if (!userId) {
    return false;
  }
  return !shouldDisplayBanner(userId);
};
