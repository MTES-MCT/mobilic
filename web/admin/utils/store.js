import React from "react";
import { useStoreSyncedWithLocalStorage } from "common/utils/store";
import flatMap from "lodash/flatMap";

const AdminStoreContext = React.createContext(() => {});

export function AdminStoreProvider({ children }) {
  const store = useStoreSyncedWithLocalStorage();

  const [users, setUsers] = React.useState([]);
  const [workDays, setWorkDays] = React.useState([]);
  const [vehicles, setVehicles] = React.useState([]);
  const [knownAddresses, setKnownAddresses] = React.useState([]);
  const [employments, setEmployments] = React.useState([]);
  const [companies, setCompanies] = React.useState([]);
  const [missions, setMissions] = React.useState([]);
  const [minWorkDaysCursor, setMinWorkDaysCursor] = React.useState(null);

  const sync = (companiesPayload, minDate) => {
    const primaryCompany = store.companies().find(c => c.isPrimary);
    setCompanies(
      companiesPayload.map(c => ({
        id: c.id,
        name: c.name,
        siren: c.siren,
        settings: c.settings,
        isPrimary: primaryCompany ? c.id === primaryCompany.id : false
      }))
    );
    setUsers(
      flatMap(
        companiesPayload.map(c => c.users.map(u => ({ ...u, companyId: c.id })))
      )
    );
    addWorkDays(companiesPayload, minDate, true);
    setVehicles(
      flatMap(
        companiesPayload.map(c =>
          c.vehicles.map(v => ({ ...v, companyId: c.id }))
        )
      )
    );
    setKnownAddresses(
      flatMap(
        companiesPayload.map(c =>
          c.knownAddresses.map(a => ({ ...a, companyId: c.id }))
        )
      )
    );
    setEmployments(
      flatMap(
        companiesPayload.map(c =>
          c.employments.map(e => ({
            ...e,
            companyId: c.id,
            company: { id: c.id, name: c.name, siren: c.siren }
          }))
        )
      )
    );
    const newMissionIds = flatMap(
      companiesPayload.map(c => c.missions.edges.map(m => m.id))
    );
    setMissions(missions => [
      ...missions.filter(m => !newMissionIds.includes(m.id)),
      ...flatMap(
        companiesPayload.map(c =>
          c.missions.edges.map(m => ({ ...m.node, companyId: c.id }))
        )
      )
    ]);
  };

  function addWorkDays(companiesPayload, minDate, reset = false) {
    let actualMinCursor = minDate;
    companiesPayload.forEach(c => {
      const wds = c.workDays;
      if (wds.pageInfo.hasNextPage) {
        const oldestWorkDay = wds.edges[wds.edges.length - 1].node;
        const oldestCursor = `${oldestWorkDay.day}${oldestWorkDay.user.id}`;
        actualMinCursor =
          oldestCursor > actualMinCursor ? oldestCursor : actualMinCursor;
      }
    });
    const actualMinDate = actualMinCursor.slice(0, 10);
    const actualMinUser = actualMinCursor.slice(10);

    const actualMaxDate =
      !reset && minWorkDaysCursor ? minWorkDaysCursor.slice(0, 10) : null;
    const actualMaxUser =
      !reset && minWorkDaysCursor ? minWorkDaysCursor.slice(10) : null;

    const workDaysToAdd = flatMap(
      companiesPayload.map(c =>
        c.workDays.edges.map(wd => ({ ...wd.node, companyId: c.id }))
      )
    ).filter(
      wd =>
        (!actualMaxDate ||
          wd.day < actualMaxDate ||
          (wd.day === actualMaxDate &&
            wd.user.id.toString() < actualMaxUser)) &&
        (wd.day > actualMinDate ||
          (wd.day === actualMinDate && wd.user.id.toString() >= actualMinUser))
    );

    setWorkDays(wds => [...(reset ? [] : wds), ...workDaysToAdd]);
    setMinWorkDaysCursor(actualMinCursor);
  }

  return (
    <AdminStoreContext.Provider
      value={{
        companies: companies,
        setCompanies,
        userId: store.userId(),
        users,
        setUsers,
        workDays,
        setWorkDays,
        vehicles,
        setVehicles,
        knownAddresses,
        setKnownAddresses,
        employments,
        setEmployments,
        missions,
        setMissions,
        sync,
        addWorkDays,
        minWorkDaysDate: minWorkDaysCursor
          ? minWorkDaysCursor.slice(0, 10)
          : null
      }}
    >
      {children}
    </AdminStoreContext.Provider>
  );
}

export const useAdminStore = () => React.useContext(AdminStoreContext);
