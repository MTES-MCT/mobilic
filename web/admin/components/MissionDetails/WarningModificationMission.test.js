import React from "react";
import { shallow } from "enzyme";
import * as api from "common/utils/api";
import { DISABLE_WARNING_MUTATION } from "common/utils/apiQueries";
import { WarningModificationMission } from "./WarningModificationMission";
import * as store from "common/store/store";
import { DISMISSABLE_WARNINGS } from "../../utils/dismissableWarnings";

let graphQlSpy;
let storeSpy;

describe("WarningModificationMission", () => {
  beforeAll(() => {
    const apiSpy = jest.spyOn(api, "useApi");
    apiSpy.mockReturnValue({
      graphQlMutate: jest.fn()
    });
    graphQlSpy = jest.spyOn(api.useApi(), "graphQlMutate");
    storeSpy = jest.spyOn(store, "useStoreSyncedWithLocalStorage");
  });

  const setup = propsOverwrite => {
    const mockUserInfo = () => {
      return { disabledWarnings: [] };
    };
    storeSpy.mockReturnValue({
      userInfo: mockUserInfo,
      ...propsOverwrite
    });
    return shallow(<WarningModificationMission />);
  };

  it("should render the WarningModificationMission correctly ", () => {
    const wrapper = setup();
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it("should not render the Warning Alert if it has already been dismissed", () => {
    const mockUserInfo = () => {
      return {
        disabledWarnings: [DISMISSABLE_WARNINGS.ADMIN_MISSION_MODIFICATION]
      };
    };
    const customUserInfo = {
      userInfo: mockUserInfo
    };
    const wrapper = setup(customUserInfo);
    expect(
      wrapper.find('[data-qa="modificationWarningCollapse"]').prop("in")
    ).toBeFalsy();
  });

  it("should render the Warning Alert if it has never been dismissed", () => {
    const wrapper = setup();
    expect(
      wrapper.find('[data-qa="modificationWarningCollapse"]').prop("in")
    ).toBeTruthy();
  });

  it("should call the dismiss warning api if the link is cliclked", () => {
    const wrapper = setup();

    const dismissWarningMessage = wrapper.find(
      '[data-qa="dismissMissionModificationWarningLink"]'
    );

    dismissWarningMessage.props().onClick({ preventDefault: () => {} });

    expect(graphQlSpy).toHaveBeenCalledWith(
      DISABLE_WARNING_MUTATION,
      { warningName: DISMISSABLE_WARNINGS.ADMIN_MISSION_MODIFICATION },
      { context: { nonPublicApi: true } }
    );
  });
});
