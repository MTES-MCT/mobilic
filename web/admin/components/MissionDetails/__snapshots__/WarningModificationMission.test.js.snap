// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WarningModificationMission should render the WarningModificationMission correctly  1`] = `
"<WithStyles(ForwardRef(Collapse)) in={true} data-qa=\\"modificationWarningCollapse\\">
  <WithStyles(ForwardRef(Alert)) severity=\\"info\\" className=\\"makeStyles-modificationAlert-3\\">
    <WithStyles(ForwardRef(List)) disablePadding={true}>
      <WithStyles(ForwardRef(ListItem)) disableGutters={true} dense={true} className=\\"makeStyles-modificationWarningItem-2\\">
        <WithStyles(ForwardRef(ListItemIcon)) className=\\"makeStyles-validationWarningIcon-1\\">
          ðŸ‘‰
        </WithStyles(ForwardRef(ListItemIcon))>
        <WithStyles(ForwardRef(ListItemText)) primary=\\"Les modifications sont enregistrÃ©es et apparaÃ®tront en cas de contrÃ´le.\\" />
      </WithStyles(ForwardRef(ListItem))>
      <WithStyles(ForwardRef(ListItem)) disableGutters={true} dense={true} className=\\"makeStyles-modificationWarningItem-2\\">
        <WithStyles(ForwardRef(ListItemIcon)) className=\\"makeStyles-validationWarningIcon-1\\">
          ðŸ‘‰
        </WithStyles(ForwardRef(ListItemIcon))>
        <WithStyles(ForwardRef(ListItemText)) primary=\\"En cas de modification des saisies, le salariÃ© recevra une notification.\\" />
      </WithStyles(ForwardRef(ListItem))>
    </WithStyles(ForwardRef(List))>
    <Link className=\\"makeStyles-dismissModificationAlert-4\\" to=\\"/#\\" data-qa=\\"dismissMissionModificationWarningLink\\" onClick={[Function: onClick]}>
      Ne plus afficher ce message
    </Link>
  </WithStyles(ForwardRef(Alert))>
</WithStyles(ForwardRef(Collapse))>"
`;
